### ISSUE TITLE
Data Products

### ISSUE THREAD

**Issue by zemogle** [LCOGT/neoexchange#539]
Increasingly in NEOx we want to be able to find files which are created as part of a process. Mostly these are images or GIFs. How do we want to interact with them and what should we store?

I propose a new model type called `DataProduct`.

Using the `ContentTypes` framework we can link each DataProduct to any model. See [Django Docs](https://docs.djangoproject.com/en/3.2/ref/contrib/contenttypes/) for more.

A basic class for `DataProduct` could be something after:

```python
class DataProduct(models.Model):
    filetype = models.SmallPositiveInteger(choices=((0,'jpeg'),(1,'fits'),(2,'gif'),(2,'csv')))
    product = models.FileField(upload_to='...')
    created = models.DateTimeField()
  # Below is the GenericForeignKey stuff from ContentTypes
    content_type = models.ForeignKey(ContentType, on_delete=models.CASCADE)
    object_id = models.PositiveIntegerField()
    content_object = GenericForeignKey('content_type', 'object_id')
```

I also propose a helper function in `core.utils` to create a `DataProduct`.

**Comment by zemogle** [comment]
File types that I think are useful are:

- jpeg
- Gif
- csv
- Image fits
- ALCDEF
- mp4
- Spectra fits

Anything I'm missing?

**Comment by talister** [comment]
> File types that I think are useful are:
> 
>     * jpeg
> 
>     * Gif
> 
>     * csv
> 
>     * Image fits
> 
>     * ALCDEF
> 
>     * mp4
> 
>     * Spectra fits
> 
> 
> Anything I'm missing?

Planetary Data System (PDS) XML label files might be one. One thing @jchate6 and I discussed was whether we wanted spelt-out seperate dataproducts for multiples of a particular type or a subtype. Considering PNG files for example we could have:

- PNGs of astrometric and photometric reference stars used
- Per-Block and combined light curves plots, both phased and unphased
- Periodogram and data window plots from a period finding code, possible several runs
- FWHM/ZP conditions plots

As one of the major goals of a `DataProducts` table as we see it, is to make an efficient "link" from "target/`Body` that produced these observations/`Blocks`" through to "these set of results and products", I think we want a way to distinguish between different flavo(u)rs of a particular product type - I don't think we would want to have a View/template just splat every PNG DataProduct it knows about for a given `Body` onto it's Body Detail page for example. Not sure what the best answer is; we seem to be proposing a richer set of DataProducts than other projects.

**Comment by zemogle** [comment]
I totally agree. This is a model which should have a way of uniquely identifying `DataProducts`. I think we have to make the model as rich as we need, so adding lots of filetypes is definitely a good thing.

If you have a look at #541 I've added more filetypes.

Currently to find an image FITS file for a body you would do the following:
```python
DataProduct.body_objects.filter(object_id=body.id, filetype=DataProduct.FITS_IMAGE)
```
This returns a queryset. The `body_objects` is `models.Manager` to simplify getting a queryset for `Body`s. There is also a `block_objects` which works the same way.

```python
DataProduct.body_objects.all()
```
will return all `DataProducts` which are attached to `Body` objects, so filtering on filetype is currently the option we have.

**Comment by jchate6** [comment]
Not everything we create makes it to S3. Do we want differentiate between files that exist only on Apophis, vs things we can expect to get from S3?

**Comment by zemogle** [comment]
The live version of NEOx should always have data products which go to S3. If you are running a local version **for testing** of NEOx, you will have configured your `MEDIA_ROOT` as a directory on your local filesystem. The `DataProducts` will still work.

If you are a naughty person, you might be tempted to point your local version at the live DB. In that case you are in the dangerous position of adding things to the live DB which will reference your local `MEDIA_ROOT`. I guess the question is... are you a naughty person?

**Comment by jchate6** [comment]
I meant that a lot of our management commands run from a chronjob on apophis. In fact MOST of the dataproducts we create are only created on apophis and never created on S3. This includes: 

- all fits files (excepting spectra)
- all of the PNG files listed
- all catalogs and LDAC files
- all spectra tarballs/FLOYDS_SPEC data products (Except final spectra fits)
- lightcurve data text files
- mpc_position text files
- ades position psv files
- light curve extraction out put lists (theoretically made defunct by new DataProducts )
- several future damit products

Currently the only files we MAKE and store on S3 are final spectra fits files and spectra guidemovies. Everything else on S3 is created on apophis and uploaded to S3, but this is a relatively minor fraction of the total number of dataproducts produced. 

All that being said, I think it would be wise to accommodate and anticipate such "naughty" behavior in some way.  Better safe than sorry.

**Comment by zemogle** [comment]
Are you saying that you routinely run code locally, but pointed at the live DB to generate dataproducts which never go to S3? If that's so, you've treating the live DB as a sort of read-only snapshot. If the files are only used by the local instance then they can't be added to the live DB.

Or are you linking to the live DB locally because the amount of data (i.e. files) is too large to upload into S3. In this case you need the computing might of Apophis to run the pipelines.

Both of these can cause problems because you have some 2 way traffic (DB access) and some 1 way (files). 

The way to solve this is with API access. Tthe pipelines would then be separated from NEOx, working on the local data but getting info from the live DB via an API. Much less "naughty" and all round safer.

**Comment by jchate6** [comment]
@zemogle one of the primary DB interactions in Neoexchange is `download_process_data` which is run from apophis on the order of 50 times a day. This code is responsible for downloading fits files, processing those fits files with astrometric and photometric pipelines, saving `frames`, linking objects, and running `lightcurve_extraction`. 
This code needs to update and/or read `block`, `body`, and `frame` (at least). Furthermore, it generates the vast majority of our data products. As I listed above, most of these are never uploaded to S3.  If this workflow (that predates me btw) is a problem, then we might want to solve that first before trying to incorporate the `DataProduct` model into a broken system.

Your continued references to "running code locally" even though I've been repeatedly trying to discuss our apophis cronjobs makes me feel like we are talking past each other a bit here. Hopefully we can resolve this at the meeting on Wednesday.

**Comment by zemogle** [comment]
By "locally" I mean something which is not available as a website on the public internet. So I am not making a distinction between Apophis and personal Django instances. If they are hitting the production DB this 

Also, what you are describing (ie. files which are never displayed in NEOx) doesn't fall into the definition of `DataProduct`. It sounds like these pipelines are generating products which never make it to the live NEOx website. This make me think this has grown organically outside of its original use case.

Hopefully a face-2-face discussion will help with all of this.

**Comment by jchate6** [comment]
Damit Data Products:
- Periodogram
- model lightcurve
- shape model
- inversion output (pole solution, sidereal period, etc) 

All of these are poorly formatted text files, but the last one is a particularly poorly formatted text file containing ~10 unlabeled numbers. I should probably re-format these data products into csv files with headers before they go live. The best solution MIGHT be to simply incorporate the inversion output into the header of the lc and shape models.

**Comment by jchate6** [comment]
From what I can see here, it looks like we won't have a copy of the ALCDEF file or the gif on apophis when everything is said and done. Direct file access to these files is useful under a number of circumstances. Furthermore, I think having a backup of our S3 files is very prudent.  Is it possible to save them to disk and then create the dataproduct and copy to default storage in the case `DATA_ROOT` != `MEDIA_ROOT`?

**Comment by zemogle** [comment]
You are correct that you won't have a local copy of ALCDEF of GIF. The reason for this is:

- You'd need unecessary logic to cope with resaving the files to a local directory
- DataProducts are available via the live website so you can get to them easily
- There is only a single place to look for these files.

I'm really not sure why having a duplicate/backup of S3 is a motivation. The database is stored there, after all. What are the other circumstances?

If there are occasions where you need look at an ALCDEF/GIF I would suggest we provide a way to easily find the web address DataProducts created via cron.

**Comment by jchate6** [comment]
Both the ALCDEF and gif files frequently need manual updating after the initial cronjob is run. We need a way to ensure that these edits overwrite the automated versions without being accidentally overwritten by an automated run in the future.

This update usually just means running the management command again with different settings, but can occasionally involve editing the text of the alcdef file by hand. Easy edit access to a version of these documents is vital. 

Furthermore, the mythical damit branch will rely on manual cron jobs that can access the ALCDEF files. If this can be done from S3 then this isn't a problem, I'm just not clear on what impact (if any) it will have to not keep a local copy.

Finally, every other data product and file created by Neoexchange has a copy on Apophis. Every fits file, every png, and up to now, every alcdef and gif. This is incredibly useful when doing actual science because the data can be easily downloaded to a local machine and worked with. It's also very useful when doing neox development, because these files are all in one place and they can be used locally without requiring a separate download from s3.

I'm certainly a big fan of allowing easy download and access to our web files from neoexchange. However, when actually using the data for science, being able to easily parse and download batches of files is important (i.e. "all ALCDEF Files for object x", or "all spectra from date1 to date2"). Will we be able to do all of these things with the proposed changes?

**Comment by zemogle** [comment]
> Both the ALCDEF and gif files frequently need manual updating after the initial cronjob is run. We need a way to ensure that these edits overwrite the automated versions without being accidentally overwritten by an automated run in the future.

Sure. We can do this with a flag setting on the dataproduct. Are there situations where a DataProduct will be overwritten?

By "manual cron jobs" do you mean a management command?

As I've said before, I'm not a fan keeping every file ever produced, in case they are useful. With `DataProducts` I'm proposing that every derived file is made a `DataProduct` not just ALCDEF and GIF. I can see the sense in downloading FITS files on Apophis as a scratch space. 

The more I hear about needing a local file to run a management command, it feels like NEOx is being made to do something which it is not suitable for. This sounds like a hack. Django is a web framework not a pipeline framework. It really feels like Apophis should be hitting API endpoints not this sort of hybrid.

> I'm certainly a big fan of allowing easy download and access to our web files from neoexchange. However, when actually using the data for science, being able to easily parse and download batches of files is important (i.e. "all ALCDEF Files for object x", or "all spectra from date1 to date2"). Will we be able to do all of these things with the proposed changes?

Really what you are talking about is more like a database query than a file look up. This is exactly one of the use cases of DataProducts.

I think you (i.e. @jchate6 and @talister ) need to decide what you want to do with this. The aim is to make your workflow stable and NEOx more functional, but at each turn we keep being pulled back to storing local files in niche circumstances.

**Comment by jchate6** [comment]
Yes, sorry about that I meant damit will probably use "manual management commands rather than cron jobs". Seems I lost some important words in the middle there.

**Comment by jchate6** [comment]
> Are there situations where a DataProduct will be overwritten?
The Gif and ALCDEF files are created automatically by regularly run cron jobs. Sometimes, they miss the target, or have contamination issues, and need to be rerun manually in this case we want the manual command to overwrite existing DataProducts. I do not believe there is any situation in which we want an automatically created DataProduct to overwrite an existing one. 

I don't think we NEED to keep a copy of these files on apophis. My point is that currently we DO, and, lacking a DataProduct model system, our current workflows rely on that. This is going to require major changes in our workflows and pipelines. We will need to develop API endpoints for these and adjust our workflow accordingly.

**Comment by talister** [comment]
> > Both the ALCDEF and gif files frequently need manual updating after the initial cronjob is run. We need a way to ensure that these edits overwrite the automated versions without being accidentally overwritten by an automated run in the future.
> 
> Sure. We can do this with a flag setting on the dataproduct. Are there situations where a DataProduct will be overwritten?
> 
> By "manual cron jobs" do you mean a management command?
> 
> As I've said before, I'm not a fan keeping every file ever produced, in case they are useful. With `DataProducts` I'm proposing that every derived file is made a `DataProduct` not just ALCDEF and GIF. I can see the sense in downloading FITS files on Apophis as a scratch space.
> 
> The more I hear about needing a local file to run a management command, it feels like NEOx is being made to do something which it is not suitable for. This sounds like a hack. Django is a web framework not a pipeline framework. It really feels like Apophis should be hitting API endpoints not this sort of hybrid.
> 
> > I'm certainly a big fan of allowing easy download and access to our web files from neoexchange. However, when actually using the data for science, being able to easily parse and download batches of files is important (i.e. "all ALCDEF Files for object x", or "all spectra from date1 to date2"). Will we be able to do all of these things with the proposed changes?
> 
> Really what you are talking about is more like a database query than a file look up. This is exactly one of the use cases of DataProducts.
> 
> I think you (i.e. @jchate6 and @talister ) need to decide what you want to do with this. The aim is to make your workflow stable and NEOx more functional, but at each turn we keep being pulled back to storing local files in niche circumstances.

I see us and the software on a longish and evolving road from the traditional "everything is local and is on One Good Server" to "everything is in the Cloud and talks through APIs" but that you are further down this road of evolving thinking than we are. When things got moved from the docknodes and into AWS, the combination of moving everything there, having to likely rapidly develop DRF APIs to be able to talk to the DB to get access to anything while analyzing data science data with old school "everything must be a file" tools and thinking was a lot to take on - when it became clearer that there could still be limited Django `Model` access direct to the DB, this made things easier and took the pressure off but also removed the impetus to continue on the road of decoupling processing tasks from `Model`-based DB access (at least it's not raw SQL statements in PHP...) 

So I think some of this "code X needs access to local files" comes from us not having moved as far down the decoupling-via-API approach and falling back to the old familiar ways of doing things. What we need is a small step back and look at it anew and realize that 'tool X needs local files in a certain format because it can't read from a S3 bucket" isn't the whole story and it's really "tool X needs to ask for these files to operate and how it gets them isn't really its problem" and can call something to get what it wants. The backend to that 'something', which is a `glob` of local files now can be swapped out for an API call for the relevant  `DataProducts` and a seamless transfer from S3 next it and the tool will be none the wiser. I think we (the SB end) just need a better refocus on "what does this thing need" and define interfaces to let it get inputs and produce outputs and then different versions of the backend providers can be written and swapped in as things evolve to the new model. So for example if we have:
`get_ALCDEF_files_for_damit(oldstyle=True)` this can have the existing `get_local_ALCDEF()` and `get_ALCDEF_from_DataProducts()`. The interface will look like:
```
def get_ALCDEF_files_for_damit(oldstyle=True):
  if oldstyle is True:
    files = get_local_ALCDEF()
  else:
    files = get_ALCDEF_from_DataProducts()
  return files
```
and then at some point soon when we're ready to make the switch, the default `kwarg` becomes `oldstyle=False` and we're into the Better New World.

**Comment by zemogle** [comment]
That sounds like a practical solution. If you are always accessing  `DataProduct` files via code, there isn't any need to care if they are local or remote. As I've said before, we should implement the `DataProduct` model soon but adapt the code in stages, starting with GIFs and ALCDEF files. As @talister suggested using wrapper functions is a good solution.

If you were to mount the Apophis file system and browse files in terminal, for example, this would be a more challenging. 

Fundamentally we all need to be on the same page about file access. If we are going to do this, we need to be constantly thinking of accessing DataProduct files via the model and not writing backdoors to get a filepath. 

I don't have the same intimate knowledge of the either the code-base or the workflow as you, @talister and @jchate6. I suggest you both checkout #541 and see what needs to done to merge this into master.

**Comment by talister** [comment]
> That sounds like a practical solution. If you are always accessing `DataProduct` files via code, there isn't any need to care if they are local or remote. As I've said before, we should implement the `DataProduct` model soon but adapt the code in stages, starting with GIFs and ALCDEF files. As @talister suggested using wrapper functions is a good solution.
> 
> If you were to mount the Apophis file system and browse files in terminal, for example, this would be a more challenging.

I think we need to evolve our thinking and practices away from "viewing local files in <image viewer of choice>" to sending those into the bucket as a `DataProduct`s and viewing them through Django.

> 
> Fundamentally we all need to be on the same page about file access. If we are going to do this, we need to be constantly thinking of accessing DataProduct files via the model and not writing backdoors to get a filepath.

Yes, definitely.

> 
> I don't have the same intimate knowledge of the either the code-base or the workflow as you, @talister and @jchate6. I suggest you both checkout #541 and see what needs to done to merge this into master.

So @jchate6 was planning to incorporate `dataproducts` into `damit` before deploy as a pathfinder and flush out any lingering issues (since he has (some) time to write code right now and I most decidedly don't with the grant); I don't *think* there will be issues in merging `dataproducts` into `damit`, followed by `dataproducts` into `master` and then `master` into `damit` (as normal before closing the `damit` PR) and `git` will work it out but squeak if this is likely to go horribly wrong (or more than normally wrong with git merges...)

**Comment by jchate6** [comment]
@talister @zemogle Are we good to close this issue?
